File: R/Regression/Position_Salary_PolynomialRegression.R
# Import Data set 
df = read.csv("C:/Users/snapp/Data-Projects/Datasets/Position_Salaries.csv")

# Remove First Column that is not necessary 
df = df[2:3]

# Split Dataset into train test split is not necessary because Dataset is too small  

# Fit Linear Regression 
LR = lm(formula = Salary ~ .,
        data = df )
summary(LR)

# Fit Polynomial Regression 
df$Level2 = df$Level^2
df$Level3 = df$Level^3
df$Level4 = df$Level^4
PLR = lm(formula = Salary ~ .,
         data = df) 
summary(PLR)

library(ggplot2)
# Visualise the Linear Regression 
ggplot() + 
  geom_point(aes(x = df$Level, y = df$Salary ),
             colour = 'red') + 
  geom_line(aes(x = df$Level, y = predict(LR, newdata = df)), 
            colour = 'blue') + 
  ggtitle('Linear Regression') +
  xlab('Level') + 
  ylab('Salary')

# Visualise the Polynomial Regression 
ggplot() + 
  geom_point(aes(x = df$Level, y = df$Salary ),
             colour = 'red') + 
  geom_line(aes(x = df$Level, y = predict(PLR, newdata = df)), 
            colour = 'blue') + 
  ggtitle('Polynomial Regression') +
  xlab('Level') + 
  ylab('Salary')

# Visualising the Polynomial Model results (for higher resolution and smoother curve)
x_grid = seq(min(df$Level), max(df$Level), 0.1)
ggplot() +
  geom_point(aes(x = df$Level, y = df$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(PLR, 
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4))),
            colour = 'blue') +
  ggtitle('Polynomial Regression') +
  xlab('Level') +
  ylab('Salary')

# Predict a New Result in Linear Regression
y_pred = predict(LR, newdata = data.frame(Level = 6.5))
# Predict a New Result in Polynomial Regression 
y_pred = predict(PLR, newdata = data.frame(Level = 6.5,
                                           Level2 = 6.5^2,
                                           Level3 = 6.5^3,
                                           Level4 = 6.5^4))

File: Python/Regression/Position_Salary_PolynomialRegression.py
# Import Libraries 
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

# Import Datasets 
df = pd.read_csv(r"C:\Users\snapp\Data-Projects\Datasets\Position_Salaries.csv")
# Independent Variables 
X = df.iloc[:,1:2].values
# X = df.iloc[:,:-1].values
# Dependent Variables 
y = df.iloc[:,-1].values

# Split Data into Train and Test is not necessary because the dataset is too small 
# Feature Scaling is not necessary beacasue the library included Feature Scaling

# Fit Linear Regression 
LR = LinearRegression()
LR.fit(X, y)

# Fit Polynomial Regression
PR = PolynomialFeatures(degree = 4)
# The model itself add one column as Intercept 
X_poly = PR.fit_transform(X)
PLR = LinearRegression()
PLR.fit(X_poly,y)

# Visualize the Linear Regression Results
plt.scatter(X, y, c = 'red', s = 10)
plt.plot(X, LR.predict(X), c = 'green')
plt.title('Linear Regression Results')
plt.xlabel("Position Level")
plt.ylabel("Salary")

# Visualize the Polynomial Regression Results
plt.scatter(X, y, c = 'red', s = 10)
plt.plot(X, PLR.predict(PR.fit_transform(X)), c = 'green')
plt.title('Polynomial Regression Results')
plt.xlabel("Position Level")
plt.ylabel("Salary")

# Visualize the Polynomial Regression Results _ Improved (Smooth Line)
X_grid = np.arange(min(X),max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid)),1)
plt.scatter(X, y, c = 'red', s = 10)
plt.plot(X_grid, PLR.predict(PR.fit_transform(X_grid)), c = 'green')
plt.title('Polynomial Regression Results')
plt.xlabel("Position Level")
plt.ylabel("Salary")

# Predict a new Result with Linear Regression
LR.predict([[6.5]])
# Predict a new Result with Linear Regression 
PLR.predict(PR.fit_transform([[6.5]]))


